src/lib/reasoners/adt_rel.ml

```diff
@@ -218,334 +327,201 @@ let assume_th_elt t th_elt _ =
-      if Options.get_debug_adt () then
-        Printer.print_dbg
-          ~module_name:"Adt_rel" ~function_name:"add_adt"
-          "new ADT expr(C): %a" E.print t;
-      { env with domains =
-                   MX.add r (HSS.singleton hs, Ex.empty) env.domains }
-
-    | _, Ty.Tadt _ ->
-      if Options.get_debug_adt () then
-        Printer.print_dbg
-          ~module_name:"Adt_rel" ~function_name:"add_adt"
-          "new ADT expr: %a" E.print t;
-      let constrs =
-        match values_of ty with
-        | None ->
-          (* The type [ty] is an ADT type. *)
-          assert false
-        | Some s -> s
-      in
-      let env =
-        { env with domains = MX.add r (constrs, Ex.empty) env.domains }
-      in
-      if HSS.cardinal constrs = 1 then
-        let h' = HSS.choose constrs in
-        let env, pending_deds =
-          deduce_is_constr uf r h' env.pending_deds env Ex.empty
-        in
-        {env with pending_deds}
-      else
-        env
-
-    | _ -> env
-
-(* Add the tester `((_ is hs) r)` to the environment [env]. *)
-let update_tester r hs env =
-  let old = try MX.find r env.seen_testers with Not_found -> HSS.empty in
-  {env with seen_testers = MX.add r (HSS.add hs old) env.seen_testers}
-
-let add env uf r t =
-  if Options.get_disable_adts () then env, []
-  else begin
-    let delayed, eqs = Rel_utils.Delayed.add env.delayed uf r t in
-    let E.{ f = sy; ty; _ } = E.term_view t in
-    let env = add_adt env uf t r sy ty in
-    { env with delayed }, eqs
-  end
+  | Ty.Tadt (name, params) ->
+    let Adt cases = Ty.type_body name params in
+    Lists.is_empty @@ Ty.assoc_destrs c cases
+  | _ -> assert false
 
-(* Update the counter of case-split size in [env]. *)
-let count_splits env la =
-  let nb =
-    List.fold_left
-      (fun nb (_,_,_,i) ->
-         match i with
-         | Th_util.CS (Th_util.Th_sum, n) -> Numbers.Q.mult nb n
-         | _ -> nb
-      )env.size_splits la
+(* Update the domains of the semantic values [r1] and [r2] according to the
+   disequality [r1 <> r2].
+
+   This function alone isn't sufficient to produce a complete decision
+   procedure for the ADT theory. For instance, let's assume we have three
+   semantic values [r1], [r2] and [r3] whose the domain is `{C1, C2}`. It's
+   clear that `(distinct r1 r2 r3)` is unsatisfiable but we haven't enough
+   information to discover this contradiction.
+
+   We plan to support model generation for ADT. As a by-product, we will got
+   a complete decision procedure for the ADT through the case-split mechanism
+   as we did in the Enum theory.
+
+   @raise Domain.Inconsistent if the disequality is inconsistent with
+          the current environment [env]. *)
+let assume_distinct ~ex r1 r2 env =
+  let d1 = Domains.get r1 env.domains in
+  let d2 = Domains.get r2 env.domains in
+  let env =
+    match Domain.as_singleton d1 with
+    | Some (c, ex1) when is_tightenable (X.type_info r2) c ->
```

Comment:
Nit: It might be simpler to keep a match on `extract r1` and `extract r2` here as previously and tighten when seeing a `Constr { c_args = []; _ }` values. It should have the same effect (propagation ensures that singleton domain become `Constr` semantic values already) while being more readable (no need for `is_tightenable`).